### DISCUSSION TITLE
bug: mutationFn being ignored

### DISCUSSION THREAD

**Discussion by Ryanjso** [trpc/trpc#4861] [Q&A]
> From @KATT:
>
> We should `Omit` `mutationFn` etc from `UseMutationOptions` (and for queries too)
>

### Provide environment information

```
  System:
    OS: macOS 13.6
    CPU: (10) arm64 Apple M1 Pro
    Memory: 123.95 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.16.1 - /usr/local/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 9.5.1 - /usr/local/bin/npm
    pnpm: 8.6.10 - /opt/homebrew/bin/pnpm
    Watchman: 2022.06.27.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 117.0.5938.92
    Safari: 16.6
  npmPackages:
    @tanstack/react-query: ^4.29.5 => 4.29.14 
    @trpc/client: ^10.21.1 => 10.23.0 
    @trpc/next: ^10.23.0 => 10.23.0 
    @trpc/react-query: ^10.21.1 => 10.23.0 
    @trpc/server: ^10.21.1 => 10.23.0 
    next: ^13.3.4 => 13.3.4 
    react: ^18.2.0 => 18.2.0 
    typescript: ^5.0.4 => 5.0.4 
 ```

### Describe the bug

`mutationFn` is not being called on `mutate.` My use-case is I need to add some variables from localStorage to my payload before it gets sent off in the mutation. `mutationFn` is never called, though `onSuccess` is.

```
export const trpc = createTRPCNext({...})
```

```
  const mutation = trpc.auth.verifyCode.useMutation({
    mutationFn: async (variables) => {
      console.log('called'); // NEVER CALLED
      const extendedVariables = {
        ...variables,
        referralCode: maybeReferralCode,
        ...validatedUtm,
      };
      return trcpProxyClient.auth.verifyCode.mutate(extendedVariables);
    },

    onSuccess(data) {
      console.log('called2'); // CALLED
      storage.setAccessToken(data.accessToken);
      storage.setRefreshToken(data.refreshToken);

      identify(data);
    },
  });

  return mutation;
};
```

### Link to reproduction

https://stackblitz.com/edit/github-ucxmx2?file=src%2Fpages%2Findex.tsx

### To reproduce

See example user mutation `updateUser`

https://stackblitz.com/edit/github-ucxmx2?file=src%2Fpages%2Findex.tsx

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4860">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4860/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/4860/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
We use the `mutationFn`  for calling tRPC so they are not usable (and shouldn't be accessible in the types)

**Comment by Ryanjso** [comment]
> We use the `mutationFn` for calling tRPC so they are not usable (and shouldn't be accessible in the types)

What if we want to mutate the payload before it is sent from the useMutation?

**Comment by Ryanjso** [comment]
Seems the only way without access to mutationFn to mutate the payload before it sends would be something like the following. Which is not the most ideal.

```
const {mutate: mutateOld, mutateAsync: mutateAsyncOld, ...rest} = trpc.updateUser.useMutation()

const mutate = (payload) => {...}
const mutateAsync = (payload) => {...}

return {mutate, mutateAsync, ...rest}
```

**Comment by KATT** [comment]
> What if we want to mutate the payload before it is sent from the useMutation?

Pass a different input to `.mutate()`

**Comment by Ryanjso** [comment]
> > What if we want to mutate the payload before it is sent from the useMutation?
> 
> Pass a different input to `.mutate()`

Yeah, though I would ideally look to avoid that because what I'm passing in is an optional param that I would need to pass in from every place I call mutate instead of one central place

**Comment by KATT** [comment]
Make a function that calls it then / a custom hook

**Comment by tim-field** [comment]
> and shouldn't be accessible in the types

They are in the types, which makes this a bit of a time sync to work out.

**Comment by Robstei** [comment]
I came across this when trying to update the query cache based on a mutation that returns a generator.

Thanks to this thread I now know that this is not possible (and might be a bad solution even it where?).

Removing queryFn and mutationFn from the types and explaining how to handle situation where you want to do something before the fetching or after the fetching but before settings the client cache would be awesome for beginners like me.
